WIP: Improve results paging

sputnick1124 opened the pull request:
I've been doing quite bit of work on trying to unify the paging interface and behavior the past couple of weekends. This is still very much a work in progress, but I would love to get some feedback from other developers if you think the current path might make things better. 

Three are a few things that are a little more complicated than I would like them to be, but the functionality is shaping up to be what I envisioned. 

~~Currently, I have only implemented paging for for the podcast episode view, but now that it is working, I should be able to implement the same for all of the library items as well pretty easily. Then I can move on to playlists and albums.~~

~~It's a bit late here right now, so I will update this description with a to-do list in the coming days to track progress. Also, I'm aware of the falling CI checks, those will be fixed before this is ready for complete review.~~

# TODO:
## Page all the things
### TrackTable (delegate track context management to the type system via type aliases?)
- [ ] MyPlaylists
- [ ] AlbumSearch
- [ ] PlaylistSearch                                                            
- [x] SavedTracks
- [ ] RecommendedTracks
- [ ] MadeForYou
### Search results
- [ ] Albums
- [ ] Artists
- [ ] Tracks
- [ ] Playlists
- [ ] Podcasts
### Library Collections
- [x] Saved artists
- [x] Saved albums
- [x] Saved tracks
- [x] Made for you
- [x] Podcasts
### Misc
- [ ] User playlists
## Other
- [ ] Library podcasts view (needs to happen before we can page results)

# Call to action
Please review the changes at any point. This PR represents a massive change to how we are handling almost every list of items in the app; I'd like to get it right so we aren't plagued with issues in the future. 

Also, let me know if there is any thing that is currently not being paged that is eligible for paging that is not captured in the to-do list above.

If anyone would like to help out, much of the work could be done in parallel from here, I think. Let me know and we can coordinate. We could either work on my fork directly, or we could target a feature branch here.

sputnick1124 committed:
stub out paging types

sputnick1124 committed:
Merge branch 'master' into improve-results-paging

Commit:
Merge branch 'master' into improve-results-paging

Commit:
WIP: paging experiment, really ugly still

sputnick1124 committed:
Move paging to its own module (*massive* commit)

Brief summary of the things contained in this commit (sorry I got in the
zone and didn't incrementally commit progress :blush:)

* Making the scrollable result type (still named
  NewScrollableResultPages to avoid collison with existing type, will
  change soon) an Option on EpisodeTable to get an easy Default was a
  bad idea. Fixed this by implementing our own Default for it.
* Moved common key event handler to associated method on paging object.
  This required the common_key_events module to become public, which I
  don't like, so I need to figure out a more elegant solution to that.
* Moved the selection index from EpisodeTable down into the paging
  object. This will be done for other objects (like TrackTable etc.)
  eventually.
* Other miscellaneous odds and ends

sputnick1124 committed:
Minor fixup to new paging interface

sputnick1124 committed:
wire in channel for ui updates

sputnick1124 committed:
make app.dispatch immutable

sputnick1124 committed:
make app.dispatch immutable

sputnick1124 committed:
send more ui window info

sputnick1124 committed:
add more things to common key handler

sputnick1124 committed:
Implement paging key handlers

This commit implements the page fetching logic for pageable lists. In
summary, the logic is to always try to have enough items in the list
ahead that we can jump a full UI page into the list. To do this, we
check the state of the list and its current index when the list is drawn
to see if wee need to fetch another page.

This requires synchronization between the pageable list and the network
IO handler since we could easily requst the same page multiple times if
we don't lock the operation. The caveat here is that each network event
handler (for paging) will have to check the lock when it is done and set
it to false to let thelist know that is safe to askfor more pages again.

sputnick1124 committed:
Merge branch 'improve-results-paging' of github.com:sputnick1124/spotify-tui into improve-results-paging

Commit:
Make library artists use new scrollable pages object

Commit:
migrate user saved albums to new ScrollableResultPages

sputnick1124 committed:
Merge branch 'master' into improve-results-paging

sputnick1124 committed:
Prevent repeat fetches of SavedArtists

sputnick1124 committed:
Clippy fixes

sputnick1124 committed:
Fix logical error in podcast shows navigation

sputnick1124 committed:
Replace errantly removed line

sputnick1124 committed:
migrate made for you playlists to new paging interface

sputnick1124 committed:
clean up paging impls a bit

sputnick1124 committed:
impl Default for Library

sputnick1124 committed:
start moving track tables into scrollable objects

sputnick1124 committed:
Rename NewScrollable... -> Scrollable... :tada:

sputnick1124 committed:
rename ui view structs, move to ui mod

Rigellute commented on src/handlers/artists.rs:
```diff
@@ -7,52 +7,21 @@ use crate::{
-        let next_index = common_key_events::on_low_press_handler(&artists.items);
-        app.artists_list_index = next_index;
-      }
+    k if common_key_events::is_list_navigation_key_event(k, app) => {
```
`is_list_navigation_key_event` looks like a great abstraction

Rigellute commented on src/paging.rs:
```diff
@@ -0,0 +1,174 @@
+      _ => unreachable!("This function cannot handle non-navigation key events!"),
+    }
+  }
+}
```
This is looking good

Rigellute commented on src/app.rs:
```diff
@@ -665,123 +662,12 @@ impl App {
-      self.library.saved_albums.index -= 1;
-    }
-  }
-
```
Will be nice to remove all this from `app`

Rigellute reviewed (commented):
This is looking promising @sputnick1124!

sputnick1124 reviewed (commented):
I think I may be noticing a degradation in performance using this fork. I'd like to hear if others are seeing the same thing.

sputnick1124 committed:
wrap various tracks in newtypes

This will allow scrollable results to continue to determine their own
dispatch based sole on the contained type. Unfortunately, this requires
(I think) adding the shim of adding items and the `next` string
separately since there is no way to dynamically cast Page<T> into
ScrollablePageResults<newtype<T>> that I know of.

sputnick1124 committed:
flesh out most of madeforyou tracks

sputnick1124 committed:
flesh out most of album tracks

took some shortcuts in the past few commits to get things working, but
need to rearrange some app locking so we don't have massive UI slowdowns
due to network activity

sputnick1124 commented:
Just a quick update for anyone that may be watching this: I have been super busy as of late and haven't been able to give this much time. I have run into a couple of design issues in implementing paged playlist tracks; I think I have an idea of an elegant solution, but haven't started implementation yet, so we'll see. Once playlist and album tracks are done, I think there won't really be any more surprises and I can zip through the rest of the checklist items.
